### ISSUE TITLE
Combining extensions

### ISSUE THREAD

**Issue by bintoro** [json-api/json-api#452]
Guys, I'm growing increasingly concerned about the extension system's ability to do what it's supposed to.

The prevailing view is that the base spec will be kept very lean, and extensions will provide all bells and whistles. It has even been pointed out that moving all optional functionality to extensions would give us feature detection for free in the form of extension negotiation.

This all sounds great until you consider that extensions can only be used one at a time.

Let's say I'm writing a public API and that there are four well-known extensions I'd like to use:
- sorting
- pagination
- sideposting
- PUT resource creation.

None of these are mutually exclusive or overlapping in functionality. In fact, I want to use the first two all the time and sometimes combine the last two as needed. Wat do?

Well, I have to define a custom extension _A_ that covers the first two and another extension called _B_ for the other half. This now means that a generic client is unable to use the full capabilities of the server even if it supports all four extensions individually.

At this point, why would anyone implement extension negotiation at all? It seems to add zero value on either the client or server. I could just as well itemize the supported functionality in the API docs, and client implementors could save the trouble of teaching their clients what _A_ and _B_ mean since no one else uses the same definitions anyway.

**TL;DR:**
- Things are directly interoperable only when you use one well-known extension at a time, but that may not be the most common scenario if everything's going to be an extension.
- There seems to be no incentive to formally declare a custom extension as opposed to just documenting the API as "JSON API + extensions this & that" (and skipping the `ext` choreography).

Maybe I'm thinking about this all wrong, so can someone please alleviate my concerns?

**Comment by hhware** [comment]
> Things are directly interoperable only when you use one well-known extension at a time, but that may not be the most common scenario if everything's going to be an extension.

Indeed, this seems to be quite an artificial limitation, if applied to arbitrary extensions, which might not overlap at all. Following that logic, things are the most directly interoperable if there are no extensions allowed! :)

Should not the spec simply tell the following:
- The value of the `ext` parameter **MUST** be a comma-separated (U+002C COMMA, ",") list of extension names.
- Servers that do not support a requested combination of extensions **MUST** return a `406 Not Acceptable` status code.

**Comment by bintoro** [comment]
@hhware, if only it were that simple. Extensions can contradict one another or have interactions that can be resolved in many equally plausible ways. It's pretty much impossible to define universal rules for dealing with those situations.

Even just criteria for determining if two extensions are compatible would be hard to come by. For example, from the point of view of a sideposting extension, it's irrelevant whether the primary resource is being created normally via POST or through another extension using PUT. But how do you formally reach that conclusion?

The only way to make multi-extension requests work is to make it each extension maintainer's responsibility to publish a list of known compatible extensions if they want to support simultaneous usage.

**Comment by hhware** [comment]
Yes, and I am suggesting a simple way of handling it: leave all this complexity outside of the spec (at least outside of this version of the spec) and allow the server to decide which combinations are supported and how to advertise that to the client. The two existing official extensions do not seem to conflict with each other, so this approach seems to be suitable for 1.0. Some non-normative note could be added to explain motivation for the current version.

**Comment by bintoro** [comment]
@hhware As I said, it's not just about outright compatibility. Two extensions might appear perfectly compatible, but different applications might have different ideas about some details (like processing order) affecting the outcome.

Therefore, "allowing the server to decide" can result in various subtly different implementations of the same extension combination. That's not good.

In practice this may turn out to be a marginal issue, but the possibility is there and must be recognized.

**Comment by hhware** [comment]
@bintoro Got it, thanks. Still, IMHO, a non-normative note would be sufficient, at least for the purposes of JSON API 1.0.

> The only way to make multi-extension requests work is to make it each extension maintainer's responsibility to publish a list of known compatible extensions if they want to support simultaneous usage.

I believe it should instead be made a responsibility of the designer of each particular API to describe extension interoperability rules for his API, if there is anything to describe in that particular case. It seems a bit unrealistic to expect that many extension maintainers would be willing to think through all possible combinations of extensions, even if the list of extensions was allowed to change only once a year.

**Comment by hhware** [comment]
Sorry, I should have written "responsibility of the designer of each particular API or library to describe extension interoperability rules for his API or library".

**Comment by dgeb** [comment]
I've been having similar concerns as we discuss possible extensions. 

The appeal of allowing multiple additive extensions has grown with me. It's easy to imagine extensions that convert a MAY in the base spec to a MUST, for instance. For instance, a "hypermedia" extension might require `self` and `related` links. Other extensions might specify support for a particular type of filtering or pagination. None of these extensions need be in conflict, and it seems artificial to merge them into one.

I agree that if we were to allow multiple extensions together, it would be the job of the server to interpret which are compatible (as @bintoro said), as well as the responsibility of the designer of each API to describe extension interoperability (as @hhware said).

And of course, there's nothing saying that any server needs to accept any particular extension or combination of extensions. Some servers will just allow one at a time.

I'm not sure this is strictly needed by 1.0 since it is an additive change, but it should be a straightforward addition if someone wants to work up a PR.

**Comment by dgeb** [comment]
Closed via #454
